modules/distributor/forwarder.go

```diff
@@ -0,0 +1,311 @@
+	t := time.NewTicker(collectInterval)
+
+	for {
+		select {
+		case <-t.C:
+			f.mutex.Lock()
+			for tenant, qm := range f.queueManagers {
+				metricForwarderQueueLength.WithLabelValues(tenant).Set(float64(qm.queueLen()))
+			}
+			f.mutex.Unlock()
+		case <-f.shutdown:
+			return
+		}
+	}
+}
+
+func (f *forwarder) start(_ context.Context) error {
+	go f.watchOverrides()
+
+	go f.watchMetrics()
+
+	return nil
+
+}
+
+func (f *forwarder) stop(_ error) error {
+	close(f.shutdown)
+	var errs []error
+	for _, tm := range f.queueManagers {
+		if err := tm.shutdown(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return multierr.Combine(errs...)
+}
+
+// queueManager manages a single tenant's queue
+type queueManager struct {
+	// wg is used to wait for the workers to drain the queue while stopping
+	wg sync.WaitGroup
+
+	tenantID         string
+	workerCount      int
+	workerAliveCount *atomic.Int32
+	queueSize        int
+	reqChan          chan *request
+	fn               forwardFunc
+	workersCloseCh   chan struct{}
+
+	readOnly *atomic.Bool
+}
+
+func newQueueManager(tenantID string, queueSize, workerCount int, fn forwardFunc) *queueManager {
+	m := &queueManager{
+		tenantID:         tenantID,
+		workerCount:      workerCount,
+		queueSize:        queueSize,
+		workerAliveCount: atomic.NewInt32(0),
+		reqChan:          make(chan *request, queueSize),
+		fn:               fn,
+		workersCloseCh:   make(chan struct{}),
+		readOnly:         atomic.NewBool(false),
+	}
+
+	m.startWorkers()
+
+	return m
+}
+
+// pushToQueue a trace to the queue
+// if the queue is full, the trace is dropped
+func (m *queueManager) pushToQueue(ctx context.Context, req *request) error {
+	if m.readOnly.Load() {
+		return fmt.Errorf("queue is read-only")
+	}
+
+	select {
+	case m.reqChan <- req:
+		// TODO: Record some metric
+	case <-ctx.Done():
+		return fmt.Errorf("failed to pushToQueue traces to tenant %s queue: %w", m.tenantID, ctx.Err())
+	default:
+		// Fail fast if the queue is full
+		return fmt.Errorf("failed to pushToQueue traces to tenant %s queue: queue is full", m.tenantID)
+
+	}
+
+	return nil
+}
+
+func (m *queueManager) startWorkers() {
+	for i := 0; i < m.workerCount; i++ {
+		m.wg.Add(1)
+		m.workerAliveCount.Inc()
+
+		go m.worker()
+	}
+}
+
+func (m *queueManager) worker() {
+	go func() {
+		defer func() {
+			m.wg.Done()
+			m.workerAliveCount.Dec()
+		}()
+
+		for {
+			select {
+			case req := <-m.reqChan:
+				// TODO: add timeout to context
+				m.forwardRequest(context.Background(), req)
+			default:
+				// Forces to always trying to pull from the queue before exiting
+				// This is important during shutdown to ensure that the queue is drained
+				select {
+				case req := <-m.reqChan:
+					m.forwardRequest(context.Background(), req)
+				case <-m.workersCloseCh:
+					// If the queue isn't empty, force to start the loop from the beginning
+					if len(m.reqChan) > 0 {
+						continue
+					}
+					return
+				}
+			}
```

Comment:
It could work with just a single select, but there is an edge case that makes it not ideal.

While the queue is being drained and the `queueManager` is being shutdown, it could happen that workers always entered the `<-workersCloseCh` case, never pulling the remaining requests from the queue.

Two selects forces that pulling from the queue is evaluated first, before the exit condition. Thus, only exiting when the queue has been empitied.